hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/AbstractClientInvocationService.java

```diff
@@ -198,34 +201,56 @@ public void run() {
                 Map.Entry<Long, ClientInvocation> entry = iter.next();
                 ClientInvocation invocation = entry.getValue();
                 ClientConnection connection = invocation.getSendConnection();
-                if (connection == null) {
-                    continue;
+                if (checkConnectionDead(connection)) {
+                    iter.remove();
+                    invocation.notifyException(newTargetDisconnectedException(connection));
+                } else if (checkHardTimeout(invocation)) {
+                    iter.remove();
+                    invocation.notifyException(newOperationTimeoutException(invocation));
                 }
 
-                if (connection.isHeartBeating()) {
-                    continue;
-                }
-
-                iter.remove();
+            }
+        }
 
-                notifyException(invocation, connection);
+        private boolean checkHardTimeout(ClientInvocation invocation) {
+            long hardTimeoutMillis = invocation.getHardTimeoutMillis();
+            if (hardTimeoutMillis == 0) {
+                return false;
             }
+            long startTimeMillis = invocation.startTimeMillis();
+            long currentTimeMillis = System.currentTimeMillis();
+            long timePassed = currentTimeMillis - startTimeMillis;
+            return timePassed > hardTimeoutMillis;
+        }
+
+        private Exception newOperationTimeoutException(ClientInvocation invocation) {
+            StringBuilder sb = new StringBuilder();
+            sb.append(invocation);
+            sb.append(" timed out because a hard timeout passed ");
+            sb.append(invocation.getHardTimeoutMillis()).append(" ms. ");
+            sb.append("Current time: ").append(timeToString(currentTimeMillis())).append(". ");
+            sb.append("Start time: ").append(timeToString(invocation.startTimeMillis())).append(". ");
+            sb.append("Total elapsed time: ").append(currentTimeMillis() - invocation.startTimeMillis()).append(" ms. ");
+            String msg = sb.toString();
+            return new OperationTimeoutException(msg);
+        }
+
+        private boolean checkConnectionDead(ClientConnection connection) {
+            return connection != null && !connection.isHeartBeating();
```

Comment:
did you forget connection==null check as it existed at current master (https://github.com/hazelcast/hazelcast/pull/12356/files#diff-163f65c5210e9a0af07867e51e5b2d6cL201) ?

or do we guarantee that it will never be null?

Reply:
If a connection is null, then this check will return `false`. In that case, we will not take any action and skip the invocation. I think, the old behaviour is preserved, right ?